### PR TITLE
nixos/corectrl: init module

### LINKED ISSUE #101764: Corectrl - create NixOS service
A "tracking issue", related to https://github.com/NixOS/nixpkgs/issues/100157

### PR DESCRIPTION
**Pull request by Artturin**
Closes #101764
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Artturin** [comment]
there's no way to run it in headless mode so we cant use systemd to autostart it

File: nixos/modules/hardware/corectrl.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+let
+  cfg = config.programs.corectrl;
+in {
+
+  options.programs.corectrl = {
+    enable = mkEnableOption ''
+      A tool to overclock amd graphics cards and processors.
+      Add your user to the corectrl group to run corectrl without it asking for your password
+    '';
+
```
Comment (AndersonTorres) [inline nixos/modules/hardware/corectrl.nix:14]:
```suggestion
```

File: nixos/modules/hardware/corectrl.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+         action.id == "org.corectrl.helperkiller.init") &&
+        subject.local == true &&
+        subject.active == true &&
+        subject.isInGroup("corectrl")) {
+            return polkit.Result.YES;
+    }
+});
+    '';
+
+    boot.kernelParams = [ "amdgpu.ppfeaturemask=0xffffffff" ];
```
Comment (AndersonTorres) [inline nixos/modules/hardware/corectrl.nix:35]: Please explain this parameter and the magical constant (in the form of a comment).

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/hardware/corectrl.nix (hunk 1 above)
Comment (Artturin) [reply nixos/modules/hardware/corectrl.nix:14]: i'll setup my editor to remove those automatically

**Comment by Artturin** [comment]
i have to do some more searching to find out whats the difference between the 2 ppfeaturemask values

File: nixos/modules/hardware/corectrl.nix [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      '';
+      ppfeaturemask = mkOption {
+        type = types.str;
+        default = "0xfffd7fff";
+        example = "0xffffffff";
+        description = ''
+          The overdrive bit for amdgpu
+          Default is 0xfffd7fff as it is less likely to cause flicker issues
+          Set this to 0xfffd7fff TODO
+          Set this to 0xffffffff TODO
```
Comment (AndersonTorres) [inline nixos/modules/hardware/corectrl.nix:27]:
```suggestion
          Sets the amdgpu.ppfeaturemask kernel option.
          In particular, it is used here to set the overdrive bit.
          Default is 0xfffd7fff as it is less likely to cause flicker issues.
```

I think it should suffice for now.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by AndersonTorres** [comment]
Many thanks in advance, @Artturin !

File: nixos/modules/hardware/corectrl.nix (hunk 3 above)
Comment (Artturin) [reply nixos/modules/hardware/corectrl.nix:27]: I think 0xffffffff should be mentioned. like this
```
...
If some feature you need isn't enabled, set this to 0xffffffff.
```

Comment (AndersonTorres) [reply nixos/modules/hardware/corectrl.nix:27]: OK, do it.

**Review by AndersonTorres — APPROVED** [review]

**Pull request merged**